### PR TITLE
Simplify winit runner exit code reporting

### PR DESCRIPTION
**Pull request by Brezak**
# Objective

~Returning a app exit code from the winit runner is complicated and deadlock prone.~
The code to return a app exit code is rather shoddy. It's use of mutex is redundant, It uses unwrap when not required and can be broken by a maintainer simply forgetting to set a value.

## Solution

Switch to using a channel.
- Deals with situations in which a event loop exits unexpectedly.
- Never panics. Even in extreme cases.

### CODE REVIEW

**Comment by Brezak** [comment]
The original plan was to switch to a arc containing a atomic u8. Although such a solution would use much less memory it would be rather annoying to deal with.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by BD103** [comment]
> Returning an app exit code from the winit runner is complicated and deadlock prone.

Is there a specific issue for this, or do you have a reproduction? I'm curious of the cause behind this change.

**Comment by BD103** [comment]
> The original plan was to switch to a arc containing a atomic u8. Although such a solution would use much less memory it would be rather annoying to deal with.

Personally, I think using atomics in this situation would not be too bad; we just use `Ordering::SeqCst` for everything and follow up by seeing if we can lower it to `Aquire` and `Release`.

Then again, I may be more familiar with concurrency.

File: crates/bevy_winit/src/lib.rs [hunk 1]
```diff
@@ -281,9 +281,7 @@ pub fn winit_runner(mut app: App) -> AppExit {
 
     let mut runner_state = WinitAppRunnerState::default();
 
-    // TODO: AppExit is effectively a u8 we could use a AtomicU8 here instead of a mutex.
-    let mut exit_status = Arc::new(Mutex::new(AppExit::Success));
-    let handle_exit_status = exit_status.clone();
+    let (exit_sender, exit_receiver) = sync_channel(1);
```
Comment (BD103) [inline crates/bevy_winit/src/lib.rs:284]:
```suggestion
    // Create a synchronous channel with a bounds of 1, since ideally only one exit code will be sent before exiting the app.
    let (exit_sender, exit_receiver) = sync_channel(1);
```

It's a bit wordy, so feel free to rephrase it!

File: crates/bevy_winit/src/lib.rs [hunk 2]
```diff
@@ -302,33 +300,32 @@ pub fn winit_runner(mut app: App) -> AppExit {
...
-            &mut focused_windows_state,
-            &mut redraw_event_reader,
-            &mut winit_events,
-            &mut exit_status,
-            event,
-            event_loop,
-        );
-    };
+    let event_handler =
+        move |event: Event<RequestRedraw>, event_loop: &EventLoopWindowTarget<UserEvent>| {
```
Comment (BD103) [inline crates/bevy_winit/src/lib.rs:304]: Can the types of these arguments be implied? I noticed in the old version `event` was implied, but `event_loop` was not.

**Review by BD103 — APPROVED** [review]
I'm not familiar with the event loop, so I recommend getting one more approval before merging.

Either way, your reasoning makes sense! I have a few small notes, but nothing blocking.

**Comment by alice-i-cecile** [comment]
@pietrosophya could you take a quick look at this if you have the time and interest?

File: crates/bevy_winit/src/lib.rs [hunk 3]
```diff
@@ -492,11 +489,6 @@ fn handle_winit_event(
                             event_loop.set_control_flow(ControlFlow::Wait);
                         }
                     }
-
-                    // Trigger the next redraw to refresh the screen immediately if waiting
-                    if let ControlFlow::Wait = event_loop.control_flow() {
-                        runner_state.redraw_requested = true;
-                    }
```
Comment (mockersf) [inline crates/bevy_winit/src/lib.rs:499]: why was this removed? it seems unrelated

Comment (Brezak) [reply crates/bevy_winit/src/lib.rs]: Accidentally removed it when I was dealing with merge conflicts. It's been put back.

**Review by mockersf — APPROVED** [review]
